### PR TITLE
LLVM_WASM: Support testing infrastructure

### PR DESCRIPTION
**Pull request by Shaikh-Ubaid**

### CODE REVIEW

**Comment by Shaikh-Ubaid** [comment]
No tests were removed from `integration_tests/CMakelists.txt`. Just `llvm_wasm` LABEL was added to supporting tests.

File: integration_tests/sin_02.f90 [hunk 1]
```diff
@@ -8,14 +8,14 @@ program sin_02
 x = 10.5_dp
 r1 = dsin(x)
 r2 = -0.87969575997167_dp
-err = abs(r1-r2)
+err = my_abs(r1-r2)
```
Comment (Shaikh-Ubaid) [inline integration_tests/sin_02.f90:11]: The name `abs` in `integration_tests/arrays_04_func_pass_arr_dims.f90` and `integration_tests/sin_02.f90` conflicts with `abs` from `math.h` when linking with c runtime library which is compiled to `wasm`. For now I simply renamed the `abs` to `my_abs` in these tests.

**Comment by Shaikh-Ubaid** [comment]
This is ready.

File: src/libasr/runtime/lfortran_intrinsics.h [hunk 2]
```diff
@@ -254,7 +254,7 @@ LFORTRAN_API int32_t _lfortran_ichar(char *c);
 LFORTRAN_API int32_t _lfortran_iachar(char *c);
 LFORTRAN_API int32_t _lfortran_all(bool *mask, int32_t n);
 LFORTRAN_API void _lpython_set_argv(int32_t argc_1, char *argv_1[]);
-LFORTRAN_API int32_t _lpython_get_argc();
+LFORTRAN_API void _lpython_get_argc(int32_t *cnt);
```
Comment (certik) [inline src/libasr/runtime/lfortran_intrinsics.h:257]: Why is this change needed?

First, we'll need to do the same change in the LLVM backend, and test it in LPython.

But since we have other functions here that return values, why cannot this one return it?

File: integration_tests/sin_02.f90 (hunk 1 above)
Comment (certik) [reply integration_tests/sin_02.f90:11]: This is problematic. If this is needed for the WASM output testing at the CI, then let's patch this file only in the WASM CI test with a patch file, and report a bug.

I would not stop testing this very important use case of a user-defined function with the same name as an intrinsic function.

File: .github/workflows/CI.yml [hunk 3]
```diff
@@ -704,6 +704,75 @@ jobs:
             ./run_tests.py -b llvm llvmImplicit
             ./run_tests.py -b llvm llvmImplicit -f
 
+  test_llvm_wasm:
+    name: Test LLVM->WASM ${{ matrix.llvm-version }}
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        llvm-version: ["10", "14", "15", "16"]
```
Comment (certik) [inline .github/workflows/CI.yml:713]: Doesn't this add a lot of tests into our CI and slow it down? Maybe we can test 10 and 16 only?

**Review by certik — COMMENTED** [review]
Beautiful, great job!

I left some comments that we should address.

**Comment by certik** [comment]
881 tests pass. Very impressive.

File: src/libasr/runtime/lfortran_intrinsics.h (hunk 2 above)
Comment (Shaikh-Ubaid) [reply src/libasr/runtime/lfortran_intrinsics.h:257]: >But since we have other functions here that return values, why cannot this one return it?

It can return a value. It can also simply take argument by reference and assign the return value to it (current approach). I think it is upto us which approach to follow.

In the LLVM backend, the prototype for this is defined as `void _lpython_get_argc(int32_t *cnt)` which conflicts with the prototype `int32_t _lpython_get_argc()`. So, we needed to update one of the prototype to match the other. In this case, I choosed to update the C runtime library prototype so that it matches with the one that we are declaring in the LLVM backend.

Comment (Shaikh-Ubaid) [reply src/libasr/runtime/lfortran_intrinsics.h:257]: Let me know if we should do the other way round.

File: .github/workflows/CI.yml (hunk 3 above)
Comment (Shaikh-Ubaid) [reply .github/workflows/CI.yml:713]: Anything is fine for me. I simply used the existing LLVM testing approach at the CI. 

>Maybe we can test 10 and 16 only?

We can. Let's also test `13` (which is in middle of 10 and 16)?

File: integration_tests/sin_02.f90 (hunk 1 above)
Comment (Shaikh-Ubaid) [reply integration_tests/sin_02.f90]: I reverted these changes.

>conflicts with abs from math.h when linking with c runtime library which is compiled to wasm

The conflict is a warning. So, the tests pass seem to pass. For now, I think it is fine to have the warning. We can later using `--mangle` with the `llvm_wasm` backend for these tests. We can also do it in this PR  as well, but since the changes in CMakeLists.txt are related to enabling a backend, I think we should provide such flags in a separate PR.

File: src/libasr/runtime/lfortran_intrinsics.h (hunk 2 above)
Comment (Shaikh-Ubaid) [reply src/libasr/runtime/lfortran_intrinsics.h]: This change breaks LPython. So, I made it the other way round by updating the LLVM backend with the correct prototype. I submitted the change to LPython as well https://github.com/lcompilers/lpython/pull/2650.

File: .github/workflows/CI.yml (hunk 3 above)
Comment (Shaikh-Ubaid) [reply .github/workflows/CI.yml]: For now, I am testing `10` and `16`. Hopefully they should be sufficient. If not we can add more later.

**Comment by Shaikh-Ubaid** [comment]
Ready.

**Review by certik — APPROVED** [review]
Looks great to merge. Thanks for this!
